### PR TITLE
Shielded folder blueprint

### PR DESCRIPTION
**Pull request by lcaggio**
This PR add a `shielded folder` blueprint. The scope of the blueprint is to create a folder implementing GCP best practices to it:
- Organizational policies
- Hirarckical firewall rules
- Cloud KMS
- VPC-SC

### CODE REVIEW

File: blueprints/data-solutions/README.md [hunk 1]
```diff
@@ -62,3 +62,10 @@ This [blueprint](./sqlserver-alwayson/) implements SQL Server Always On Availabi
 This [blueprint](./vertex-mlops/) implements the infrastructure required to have a fully functional MLOPs environment using Vertex AI: required GCP services activation, Vertex Workbench, GCS buckets to host Vertex AI and Cloud Build artifacts, Artifact Registry docker repository to host custom images, required service accounts, networking and Workload Identity Federation Provider for Github integration (optional).
 
 <br clear="left">
+
+### Shielded Folder
+
+<a href="./shielded-folder/" title="Shielded Folder"><img src="./shielded-folder/images/overview_diagram.png" align="left" width="280px"></a>
+This [blueprint](./shielded-folder/) implements an opinionated Folder configuration to implement GCP best practices. Configurations implemented on the folder would be beneficial to host Workloads inheriting constraints from the folder they belong to.
```
Comment (ludoo) [inline blueprints/data-solutions/README.md:69]: implements... implement
maybe `according to GCP best practices`?
Folder lowercase BTW

Comment (ludoo) [reply blueprints/data-solutions/README.md:69]: Configuration `set` (or `defined`) on the folder....
lowercase workloads

File: blueprints/data-solutions/shielded-folder/README.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
+# Shielded folder
+
+This blueprint implements an opinionated Folder configuration to implement GCP best practices. Configurations implemented on the folder would be beneficial to host Workloads inheriting constraints from the folder they belong to.
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:3]: same comment as the higher level README

File: blueprints/data-solutions/shielded-folder/README.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+In this blueprint, a folder will be created implementing the following features:
+- Organizational policies
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:6]: newline before lists

tip: install a Markdown validator in vscode

File: blueprints/data-solutions/shielded-folder/README.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+- Hierarchical firewall rules
+- Cloud KMS
+- VPC-SC
+
+Within the folder, the following projects will be created:
+- 'audit-logs' where Audit Logs sink will be created
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:12]: newline before lists

File: blueprints/data-solutions/shielded-folder/README.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+The following diagram is a high-level reference of the resources created and managed here:
+
+![Shielded architecture overview](./images/overview_diagram.png "Shielded architecture overview")
+
+## Design overview and choices
+
+Despite its simplicity, this blueprint implements the basics of a design that we've seen working well for various customers.
+
+The approach adapts to different high-level requirements:
+- IAM roles inheritance
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:25]: newline before lists

File: blueprints/data-solutions/shielded-folder/README.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+- Organizational policies
+- Audit log sink
+- VPC Service Control
+- Cloud KMS
+
+## Project structure
+The Shielded Folder blueprint is designed to rely on several projects:
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:32]: newline after heading and before list

File: blueprints/data-solutions/shielded-folder/README.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+- `audit-log`: to host Audit logging buckets and Audit log sync to GCS, BigQuery or PubSub
+- `sec-core`: to host security-related resources such as Cloud KMS and Cloud Secrets Manager
+
+This separation into projects allows adhering to the least-privilege principle by using project-level roles.
+
+## User groups
+User groups provide a stable frame of reference that allows decoupling the final set of permissions from the stage where entities and resources are created, and their IAM bindings are defined.
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:39]: newline after heading

File: blueprints/data-solutions/shielded-folder/README.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+We use three groups to control access to resources:
+- `data-engineers`: They handle and run workloads on the `workload` subfolder. They have editor access to all resources in the `workload` folder in order to troubleshoot possible issues within the workload. This team can also impersonate any service account in the workload folder.
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:42]: newline before lists, and I only see two groups mentioned not three

File: blueprints/data-solutions/shielded-folder/README.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+- `data-security`: They handle security configurations for the shielded folder. They have owner access to the `audit-log` and `sec-core` projects.
+
+## Encryption
+The blueprint supports the configuration of an instance of Cloud KMS to handle encryption on the resources. The encryption is disabled by default, but you can enable it by configuring the `enable_features.encryption` variable.
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:46]: newline after heading

File: blueprints/data-solutions/shielded-folder/README.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+The script will create keys to encrypt log sink buckets/datasets/topics in the specified regions. Configuring the `kms_keys` variable, you can create additional KMS keys needed by your workload.
+
+## Customizations
+
+### Organization policy
+You can configure the Organization policy enforced on the folder editing yaml files in the [org-policies](./data/org-policies/) folder. An opinionated list of policies that we suggest enforcing is listed.
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:53]: newline

Comment (ludoo) [reply blueprints/data-solutions/shielded-folder/README.md:53]: the organization `policies`

File: blueprints/data-solutions/shielded-folder/README.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+Some additional Organization policy constraints you may want to evaluate adding:
+* 'constraints/gcp.resourceLocations': to define the locations where location-based GCP resources can be created.
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:56]: `-` instead of `*` and missing newline

Comment (ludoo) [reply blueprints/data-solutions/shielded-folder/README.md:56]: please also use backticks for code formatting on policy names

File: blueprints/data-solutions/shielded-folder/README.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+* 'constraints/gcp.restrictCmekCryptoKeyProjects': to define which projects may be used to supply Customer-Managed Encryption Keys (CMEK) when creating resources.
+
+### VPC Service Control
+VPC Service Control is configured to have a Perimeter containing all projects within the folder. Additional projects you may add to the folder won't be automatically added to the perimeter, and a new `terraform apply` is needed to add the project to the perimeter.
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:60]: newline

File: blueprints/data-solutions/shielded-folder/README.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+The VPC SC configuration is set to dry-run mode, but switching to enforced mode is a simple operation involving modifying a few lines of code highlighted by ad-hoc comments.
+
+Access level rules are not defined. Before moving the configuration to enforced mode, configure access policies to continue accessing resources from outside of the perimeter. 
+
+An access level based on the network range you are using to reach the console (e.g. Proxy IP, Internet connection, ...) is suggested. Example:
+
+```
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:68]: missing hcl specification after backticks

File: blueprints/data-solutions/shielded-folder/README.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    conditions = [
+      { members = ["user:user1@example.com"] }
+    ]
+  }
+}
+```
+
+Alternatively, you can configure an access level based on the identity that needs to reach resources from outside the perimeter.
+
+```
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:80]: missing hcl

File: blueprints/data-solutions/shielded-folder/README.md [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+  users = {
+  conditions = [
+        { ip_subnetworks = ["101.101.101.0/24"] }
+      ]
+  }
+}
+```
+
+## How to run this script
+To deploy this blueprint in your GCP organization, you will need
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:91]: newline before and after

File: blueprints/data-solutions/shielded-folder/README.md [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+- a folder or organization where resources will be created
+- a billing account that will be associated with the new projects
+
+The Shielded Folder blueprint is meant to be executed by a Service Account (or a regular user) having this minimal set of permission:
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:95]: newline before list

File: blueprints/data-solutions/shielded-folder/README.md [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+- Billing account
+  - `roles/billing.user`
+- Folder level
+  - `roles/resourcemanager.folderAdmin`
+  - `roles/resourcemanager.projectCreator`
+
+The shielded Folfer blueprint assumes [groups described](#groups) are created in your GCP organization.
+
+### Variable configuration
+There are three sets of variables you will need to fill in:
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:105]: newline before and after

File: blueprints/data-solutions/shielded-folder/README.md [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+```
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:106]: hcl

File: blueprints/data-solutions/shielded-folder/README.md [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+organization = {
+  id     = "12345678"
+  domain = "example.com"
+}
+prefix = "prefix"
+```
+
+### Deploying the blueprint
+Once the configuration is complete, run the project factory by running
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/README.md:115]: newline

**Review by ludoo — COMMENTED** [review]
dropped a few comments in Markdown syntax, will tackle Terraform files next

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# tfdoc:file:description Variables definition.
+
+variable "access_policy" {
+  description = "Access Policy name, set to null if creating one."
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/variables.tf:18]: can you merge this with `access_policy_create` and combine into an `access_policy_config` variable?

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+variable "data_dir" {
+  description = "Relative path for the folder storing configuration data."
+  type        = string
+  default     = "data"
+}
+
+variable "enable_features" {
+  description = "Flag to enable features on the solution."
+  type = object({
+    encryption = bool
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/variables.tf:42]: can you use `optional` here so users can override a single value?

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+    log_sink   = bool
+    vpc_sc     = bool
+  })
+  default = {
+    encryption = false
+    log_sink   = true
+    vpc_sc     = true
+  }
+}
+variable "folder_create" {
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/variables.tf:52]: same as for access policy, I would combine in `folder_config`

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+  description = "Folder ID in case you use folder_create=null."
+  type        = string
+  default     = null
+}
+
+variable "groups" {
+  description = "User groups."
+  type        = map(string)
+  default = {
+    workload-engineers = "gcp-data-engineers"
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/variables.tf:71]: please use a strict type and optional with defaults here to allow overriding individual values

```hcl
type = object({
  workload-engineers = optional(string, "gcp-data-engineers")
  workload-security = optional(string, "gcp-data-security")
})
default = {}
nullable = false
```

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+  type = object({
+    domain = string
+    id     = string
+  })
+}
+
+variable "prefix" {
+  description = "Prefix used for resources that need unique names. Use 9 characters or less."
+  type        = string
+
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/variables.tf:140]: extra newline

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+  validation {
+    condition     = try(length(var.prefix), 0) < 10
+    error_message = "Use a maximum of 9 characters for prefix."
+  }
+}
+
+variable "projects_create" {
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/variables.tf:147]: can you combine this and project ids into a single variable, as per comments above?

**Review by ludoo — COMMENTED** [review]
added a few comments on variables

File: tests/blueprints/data_solutions/shielded-folder/simple.tfvars [hunk 26]
```diff
@@ -0,0 +1,13 @@
+data_dir = null
```
Comment (juliocc) [inline tests/blueprints/data_solutions/shielded-folder/simple.tfvars:1]: Remember to remove the dash from this folder

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 27]
```diff
@@ -0,0 +1,225 @@
...
+  type = object({
+    domain = string
+    id     = string
+  })
+}
+
+variable "prefix" {
+  description = "Prefix used for resources that need unique names. Use 9 characters or less."
+  type        = string
+  validation {
```
Comment (juliocc) [inline blueprints/data-solutions/shielded-folder/variables.tf:135]: This makes sense for fast but I don't think we need it here.

File: blueprints/data-solutions/shielded-folder/variables.tf [hunk 28]
```diff
@@ -0,0 +1,221 @@
+# Copyright 2023 Google LLC
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/variables.tf:1]: I would add validation rules where a variable has two or more optional attributes, but at least one should be passed. Feel free to resolve this and it after merging :)

File: blueprints/data-solutions/shielded-folder/log-export.tf [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+locals {
+  gcs_storage_class = (
+    length(split("-", var.log_locations.storage)) < 2
+    ? "MULTI_REGIONAL"
+    : "REGIONAL"
+  )
+  log_types = toset([for k, v in var.log_sinks : v.type])
+
+  _log_keys = var.enable_features.encryption ? {
+    bq      = var.enable_features.log_sink ? ["projects/${module.sec-project.0.project_id}/locations/${var.log_locations.bq}/keyRings/${var.log_locations.bq}/cryptoKeys/bq"] : null
```
Comment (ludoo) [inline blueprints/data-solutions/shielded-folder/log-export.tf:28]: we need to find a way of breaking these long lines, I can try after you merge

**Review by ludoo — APPROVED** [review]

File: blueprints/data-solutions/shielded-folder/variables.tf (hunk 28 above)
Comment (lcaggio) [reply blueprints/data-solutions/shielded-folder/variables.tf:1]: added on 'access_policy_config' and 'project_config'

File: blueprints/data-solutions/shielded-folder/log-export.tf (hunk 29 above)
Comment (lcaggio) [reply blueprints/data-solutions/shielded-folder/log-export.tf:28]: Thanks

**Pull request closed without merging**